### PR TITLE
Update PutFile.java: fix path traversal vulnerability

### PR DESCRIPTION
**Pull request by azgron**
Check if filename contains path traversal

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by joewitt** [comment]
For things which are vulnerabilities the proper handling process is documented here: https://nifi.apache.org/security.html

For this item I am not comfortable with us calling this a vulnerability. The processor by intent allows a NiFi process running as a particular OS user to perform a recursive listing including following symlinks to authorized resources.  If another user with access to that location includes a link which goes beyond that directory this is a reasonable thing.  We could add a property to attempt to prevent such a thing but it would be purely based on managing user intent/expecation/control.  It isn't about a vulnerability.

**Comment by joewitt** [comment]
Sorry to clarify this is not on recursive listing...this is on file placement.  The user nifi runs as will attempt to resolve the directory location specified in flowfile metadata with the actual authorized path on the system.  There are certainly cases where an admin would want such control.  I'd be fine with an option to prevent writing to a different location but we need to not just do it as a default thing.

**Comment by MikeThomsen** [comment]
I agree with @joewitt. I wouldn't call this a vulnerability unless you're running NiFi as root or something (and let's be honest, this is the least of your worries if that's the case).

**Comment by azgron** [comment]
Thanks for your comments.

Do you agree to add a boolean property witch called "AvoidPathEscapes"?
The default value will be false as @joewitt mentioned.

**Comment by joewitt** [comment]
Yep I am fine with that - perhaps 'prevent' is better than 'avoid'  The display name would be 'Prevent Path Escape' and the name would be 'preventpathescape' or something consistent with how others do it.  Default of false is good.  Description will let the user know the purpose of the property if true is to detect whether the resolved path (including following symlinks) still appears inline with the intended specified target directory.   Where this could be confusing is when the base dir itself is a symlink and resolves to something else.  But the more context you can give the user on the intent and so long as this is purely optional it is fine.

**Comment by azgron** [comment]
I have done the changes. What do you think?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java [hunk 1]
```diff
@@ -232,8 +240,14 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             final Path rootDirPath = configuredRootDirPath.toAbsolutePath();
             String filename = flowFile.getAttribute(CoreAttributes.FILENAME.key());
             final Path tempCopyFile = rootDirPath.resolve("." + filename);
-            final Path copyFile = rootDirPath.resolve(filename);
-
+            final Path copyFile = rootDirPath.resolve(filename)
+            if (context.getProperty(PREVENT_PATH_ESCAPE).asBoolean() && !copyFile.startsWith(rootDirPath)) {
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                logger.error("Resolved path escapes the root dir path");
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java:247]: I would provide the two path values that were compared/tested so the user can understand what really happened.

Comment (azgron) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java:247]: Do you mean:
!(copyFile.startsWith(rootDirPath) ||  tempCopyFile.startsWith(rootDirPath))

**Comment by azgron** [comment]
@joewitt @MikeThomsen What do you think? What are the next steps?

**Comment by exceptionfactory** [comment]
@azgron The current commit includes a checkstyle violation on line 250, causing failures on Linux and MacOS.  The PR checklist includes several steps that will help streamline the review process, including prefixing the PR and commit messages with an associated Jira issue number, adding unit tests, and including a displayName() setting on new properties.  Based on previous feedback, it may be better to close this PR, create a Jira issue, and then create a new PR after working through the applicable checklist steps.

**Pull request closed without merging**